qutebrowser/javascript/webengine_caret.js

```diff
@@ -0,0 +1,1104 @@
+                        }
+                    }
+                    if (siblingNode) {
+                        cursor.node = siblingNode;
+                        cursor.text = TraverseUtil.getNodeText(siblingNode);
+                        cursor.index = 0;
+
+                        if (cursor.node.constructor !== Text) {
+                            nodesCrossed.push(cursor.node);
+                        }
+
+                        break;
+                    }
+
+                    const parentNode = cursor.node.parentNode;
+                    if (parentNode &&
+                        parentNode.constructor !== HTMLBodyElement) {
+                        cursor.node = cursor.node.parentNode;
+                        cursor.text = null;
+                        cursor.index = 0;
+                    } else {
+                        return null;
+                    }
+                }
+            }
+        }
+    };
+
+    TraverseUtil.getNextChar = function( // eslint-disable-line max-params
+        startCursor, endCursor, nodesCrossed, skipWhitespace) {
+        startCursor.copyFrom(endCursor);
+        let fChar = TraverseUtil.forwardsChar(endCursor, nodesCrossed);
+        if (fChar === null) {
+            return null;
+        }
+
+        const initialWhitespace = TraverseUtil.isWhitespace(fChar);
+
+        while ((TraverseUtil.isWhitespace(fChar)) ||
+            (TraverseUtil.isSkipped(endCursor.node))) {
+            fChar = TraverseUtil.forwardsChar(endCursor, nodesCrossed);
+            if (fChar === null) {
+                return null;
+            }
+        }
+        if (skipWhitespace || !initialWhitespace) {
+            startCursor.copyFrom(endCursor);
+            startCursor.index--;
+            return fChar;
+        }
+
+        for (let i = 0; i < nodesCrossed.length; i++) {
+            if (TraverseUtil.isSkipped(nodesCrossed[i])) {
+                endCursor.index--;
+                startCursor.copyFrom(endCursor);
+                startCursor.index--;
+                return " ";
+            }
+        }
+        endCursor.index--;
+        return " ";
+    };
+
+    const CaretBrowsing = {};
+
+    CaretBrowsing.isEnabled = false;
+
+    CaretBrowsing.forceEnabled = false;
+
+    CaretBrowsing.onEnable = undefined;
+
+    CaretBrowsing.onJump = undefined;
+
+    CaretBrowsing.isWindowFocused = false;
+
+    CaretBrowsing.isCaretVisible = false;
+
+    CaretBrowsing.caretElement = undefined;
+
+    CaretBrowsing.caretX = 0;
+
+    CaretBrowsing.caretY = 0;
+
+    CaretBrowsing.caretWidth = 0;
+
+    CaretBrowsing.caretHeight = 0;
+
+    CaretBrowsing.caretForeground = "#000";
+
+    CaretBrowsing.caretBackground = "#fff";
+
+    CaretBrowsing.isSelectionCollapsed = false;
+
+    CaretBrowsing.blinkFunctionId = null;
+
+    CaretBrowsing.targetX = null;
+
+    CaretBrowsing.blinkFlag = true;
+
+    CaretBrowsing.isWindows =
+        window.navigator.userAgent.indexOf("Windows") !== -1;
```

Comment:
This will break when the user customizes the UA, most likely? Can't we format this information into the file when reading it with Python instead?

Reply:
Agreed.